src/legacy/reduce.rs

```diff
@@ -0,0 +1,486 @@
+use std::collections::VecDeque;
+use std::io::{self, copy, Read, Result};
+
+use thiserror::Error;
+
+use crate::legacy::lz77::lz77_output_backref;
+
+use super::bitstream::{lsb, BitStream};
+
+#[derive(Error, Debug)]
+enum ReduceError {
+    #[error("Invalid follower set")]
+    InvalidFollowerSet,
+    #[error("Error reading next byte")]
+    ErrorReadingNextByte,
+}
+
+/// Number of bits used to represent indices in a follower set of size n.
+fn follower_idx_bw(n: u8) -> u8 {
+    assert!(n <= 32);
+
+    if n > 16 {
+        return 5;
+    }
+    if n > 8 {
+        return 4;
+    }
+    if n > 4 {
+        return 3;
+    }
+    if n > 2 {
+        return 2;
+    }
+    if n > 0 {
+        return 1;
+    }
+    return 0;
+}
+
+#[derive(Default, Clone, Copy)]
+struct FollowerSet {
+    size: u8,
+    idx_bw: u8,
+    followers: [u8; 32],
+}
+
+/// Read the follower sets from is into fsets. Returns true on success.
+fn read_follower_sets(is: &mut BitStream, fsets: &mut [FollowerSet]) -> bool {
+    for i in (0..=255 as usize).rev() {
+        let n = lsb(is.bits(), 6) as u8;
+        if n > 32 {
+            return false;
+        }
+        if !is.advance(6) {
+            return false;
+        }
+        fsets[i].size = n;
+        fsets[i].idx_bw = follower_idx_bw(n);
+
+        for j in 0..fsets[i].size as usize {
+            fsets[i].followers[j] = is.bits() as u8;
+            if !is.advance(8) {
+                return false;
+            }
+        }
+    }
+
+    return true;
+}
+
+/// Read the next byte from is, decoded based on prev_byte and the follower sets.
+/// The byte is returned in *out_byte. The function returns true on success,
+/// and false on bad data or end of input.
+fn read_next_byte(
```

Comment:
Refactor this to return an `Option<u8>`. Compilers optimize functions better when they return values and don't mutate more references than necessary.